modules/raft/src/main/java/org/apache/ignite/raft/jraft/rpc/impl/ActionRequestProcessor.java

```diff
@@ -82,25 +84,30 @@ public void handleRequest(RpcContext rpcCtx, ActionRequest request) {
 
         JraftServerImpl.DelegatingStateMachine fsm = (JraftServerImpl.DelegatingStateMachine) node.getOptions().getFsm();
 
-        if (request.command() instanceof WriteCommand) {
+        if (request instanceof WriteActionRequest) {
             if (fsm.getListener() instanceof BeforeApplyHandler) {
                 synchronized (groupIdSyncMonitor(request.groupId())) {
                     callOnBeforeApply(request, fsm);
-                    applyWrite(node, request, rpcCtx);
+                    applyWrite(node, (WriteActionRequest) request, rpcCtx);
                 }
             } else {
-                applyWrite(node, request, rpcCtx);
+                applyWrite(node, (WriteActionRequest) request, rpcCtx);
             }
         } else {
             if (fsm.getListener() instanceof BeforeApplyHandler) {
                 callOnBeforeApply(request, fsm);
             }
 
-            applyRead(node, request, rpcCtx);
+            applyRead(node, (ReadActionRequest) request, rpcCtx);
         }
     }
+
     private static void callOnBeforeApply(ActionRequest request, DelegatingStateMachine fsm) {
-        ((BeforeApplyHandler) fsm.getListener()).onBeforeApply(request.command());
+        Command command = request instanceof WriteActionRequest
+                ? ((WriteActionRequest) request).command()
+                : ((ReadActionRequest) request).command();
```

Comment:
Can we create one common interface that will return a command so as not to use `instanceof`?

Reply:
No, this is deliberate

Reply:
ok